core/js/decanter.js

```diff
@@ -1,3 +1,434 @@
+     * @param {HTMLElement|Nav} nav - the Nav that contains the element. May be a main nav (<nav>) or a subnav (Nav)
+     */
+    constructor( item, nav ) {
+      this.item = item;
+      this.nav = nav;
+      this.link = this.item.querySelector( 'a' );
+      this.subNav = null;
+
+      this.item.addEventListener( 'keydown', this );
+
+      if ( this.isSubNavTrigger() ) {
+        this.subNav = new Nav( this );
+        theSubNavs.push( this ); // maintain globsl list of subnavs for closeAllSubNavs()
+        this.item.addEventListener( 'click', this );
+      }
+    }
+
+    isFirstItem() { return this.nav.items.indexOf( this ) === 0; }
+    isLastItem() { return this.nav.items.indexOf( this ) === this.nav.items.length - 1; }
+    isSubNavTrigger() { return this.item.lastElementChild.tagName.toUpperCase() === 'UL'; };
+    isSubNavItem() { return this.isSubNavTrigger() || this.nav.isSubNav(); }
+    isExpanded() { return this.link.getAttribute('aria-expanded') === 'true'; }
+    setExpanded( value ) { this.link.setAttribute( 'aria-expanded', value ); }
+
+    openSubNav() {
+      closeAllSubNavs();
+
+      if ( this.isSubNavTrigger() ) {
+        this.item.classList.add( 'su-main-nav__item--expanded' );
+        this.setExpanded( 'true' );
+        this.subNav.focusOn( 'first' );
+      }
+    };
+
+    closeSubNav( focusOnTrigger = false ) {
+      if ( this.isSubNavTrigger() ) {
+        this.item.classList.remove( 'su-main-nav__item--expanded' );
+        this.setExpanded( 'false' );
+        if ( focusOnTrigger ) this.link.focus();
+      }
+      else if ( this.isSubNavItem() ) {
+        this.nav.elem.closeSubNav( focusOnTrigger ); // this.nav.elem should be a subNavTrigger
+      }
+    };
+
+    //////
+    // Event handlers
+
+    /**
+     * Handler for all events attached to an instance of this class. This method must exist when events are bound
+     * to an instance of a class (vs a function or method). This method is called for all events bound to an instance.
+     * It inspects the instance (this) for an appropriate handler based on the event type. If found, it dispatches
+     * the event to the appropriate handler.
+     *
+     * @param {KeyboardEvent} event
+     * @returns {*}
+     */
+    handleEvent( event ) {
+      event = event || window.event;
+      // if this class has an onEvent method, e.g. onClick, onKeydown, invoke it
+      const handler = 'on' + event.type.charAt(0).toUpperCase() + event.type.slice(1);
+      if ( typeof  this[ handler ] === 'function' ) {
+        const target = event.target || event.srcElement; // the element that was clicked
+        return this[ handler ]( event, target );
+      }
+    }
+
+    /**
+     * Handler for keydown events. keydown is bound to all NavItem's.
+     * Dispatched from this.handleEvent().
+     *
+     * @param {KeyboardEvent} event
+     * @param {HTMLElement} target
+     */
+    onKeydown( event, target ) {
```

Comment:
The cyclomatic complexity of this function is 21, which is high. It would be nice if this could be refactored a bit to reduce the overall complexity and increase the readability.

Reply:
It is somewhat complicated processing - there are several special cases.

Nothing is leaping to mind as to how to make this more readable. The only thing I can think of is to move the code that differs on desktop vs. mobile, i.e. the arrow handling, to separate functions like, i dunno, maybe `onDesktopKeydown()` and `onMobileKeydown()`. That might reduce the numerical complexity, but I'm not sure it would aid in human readability.

Do you have any specific suggestions on what would improve readability?

Reply:
Agreed, this one is tough as there are a lot of conditions to be bet with one event listener. 

Perhaps refactoring the if/else block into a switch statement and move the conditional body into their own methods. eg:
```
 switch( theKey ) {
  case 'home':
  case 122:
    this.homeKeyPress();
    break;
  
  case 'End':
  case 123:
    this.endKeyPress();
    break;
...
}
```